Fix sporadically failing fuse tests

evgmik opened the pull request:
Fixes #78
The fuse tests were sporadically failing often leaving <defunct>
processes which were loading process queue. I observed load of 30
for several our until reboot, though there we no negative effects.

After investigation, we found that this is due to `go` (v1.9 and up)
ways of handling poll requests. Here are the outcomes:

The routines which manage fuse layer
and OS dependent code (os.Open, and similar)
MUST BE SEPARATE OS PROCESSES.  Note: not different `go` routines but processes!
See https://github.com/bazil/fuse/issues/264#issuecomment-727269770
This separation happens automatically during normal brig operations, but
TESTING FUSE LAYER IN GO IS VERY TRICKY.
See brig relevant discussion at
https://github.com/sahib/brig/pull/77#issuecomment-754831080
However, this problem is general for any go program from version 1.9,
as can be seen in references to the issue.

bazil/fuse offers "bazil.org/fuse/fs/fstestutil/spawntest"
infrastructure which helps run tests in different communicating via socket processes.

This commit rewrites fuse test to use methods from `spawntest`.

evgmik committed:
Fix sporadically failing fuse tests

Fixes #78
The fuse tests were sporadically failing often leaving <defunct>
processes which were loading process queue. I observed load of 30
for several our until reboot, though there we no negative effects.

After investigation, we found that this is due to `go` (v1.9 and up)
ways of handling poll requests. Here are the outcomes:

The routines which manage fuse layer
and OS dependent code (os.Open, and similar)
MUST BE SEPARATE OS PROCESSES.  Note: not different `go` routines but processes!
See https://github.com/bazil/fuse/issues/264#issuecomment-727269770
This separation happens automatically during normal brig operations, but
TESTING FUSE LAYER IN GO IS VERY TRICKY.
See brig relevant discussion at
https://github.com/sahib/brig/pull/77#issuecomment-754831080
However, this problem is general for any go program from version 1.9,
as can be seen in references to the issue.

bazil/fuse offers "bazil.org/fuse/fs/fstestutil/spawntest"
infrastructure which helps run tests in different communicating via socket processes.

This commit rewrites fuse test to use methods from `spawntest`.

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
-	// Try to read it over fuse:
-	helloBuffer := &bytes.Buffer{}
-	fd, err := os.Open(path)
+func withMount(t *testing.T, opts MountOptions, f func(ctx context.Context, control *spawntest.Control, mount *mountInfo)) {
```
I saw that you often pass the zero value of MountOptions to ``withMount()``. It's common practice to make ``opts`` a pointer and fill the defaults in ``withMounts`` if it's nil. Just a small nitpick, but makes code a bit tidier I think.

evgmik replied:
I would like to keep it this way. Otherwise I had to constantly check the definition of the function to see what place the nil is taking.

My way is good for copy-pasting a test header, I see exactly what is sent to withMount. It is more typing if you do it from scratch, but who does it?

Am I missing something?

sahib replied:
I'm not opposed to leaving it as it is. I just took it as an example for a typical Go-technique. So I will approve both ways - it's more of a philosophical question after all.

> Otherwise I had to constantly check the definition of the function to see what place the nil is taking.

Fair point, but usually good text editors IDEs help with that. If I don't remember what e.g. the third parameter I just jump to the function definition. If I always have to do it, it's usually a sign for me to simplify the function because it might have too many parameters. Arguably languages like python make this easier by the optional use of keyword arguments like ``f(buffer_size=4)``. 

> ...but who does it?

Me often, when I'm unfamiliar with some code, I try to avoid copy&pasting. Typing function calls makes you think. Later on I do plenty of copy&paste too, though. Anyway, APIs should be designed to be easy to be used and understood, not to be easy to copy & paste.

evgmik replied:
I left it as it were. But it is time for me to tune up my Vim

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
-		withMountFromFs(t, opts, fs, f)
-	})
+var helpers spawntest.Registry
+var brigHelper = helpers.Register("brigHelp", &brigHelp{})
```
Prefixing names with the program name is rather unusual in Go, I suggest to use more specific names like ``fuseHelper`` or something.

evgmik replied:
My bad: `brigFS` is officially `catfs`. But I did this mental substition through the `fuse-test` code.  Why do you call it `cat...`, is it short for catalog?

But I see why it would be confusing, for those who read the code. I will relabel the problem spots.

sahib replied:
It stands for **c**ontent **a**ddressed **t**ransactional **f**ile**s**ystem. But that's just alibi, I just like cats.

evgmik replied:
addressed with massive renaming in f33b437

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
-		t.Fatalf("Unable to open simple file over fuse: %v", err)
+		t.Fatalf("Failed to create temp dir for brig file system: %v", err)
 	}
+	require.Nil(t, err)
```
Error double check?

evgmik replied:
fixed in 5a940b7

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
+	if err := control.JSON("/mount").Call(ctx, req, &nothing{}); err != nil {
+		t.Fatalf("calling helper: %v", err)
 	}
+	require.Nil(t, err)
```
Error double check?

evgmik replied:
fixed in 5a940b7

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
+	f(ctx, control, &mount)
+
+	// cleanup and unmount
+	if err := control.JSON("/unmount").Call(ctx, nothing{}, &nothing{}); err != nil {
```
It might be an idea to execute this code in a defer func, so it gets executed always, even if the test panics. Otherwise we might still have a (dysfunctional) mount lying around in that case.

evgmik replied:
done in 34b7988

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
-			len(data),
-			helloBuffer.Len(),
-		)
+func checkFuseFileMatcheToBrig(t *testing.T, ctx context.Context, control *spawntest.Control, fusePath string, brigPath string) {
```
Small typo: ``Matche``.

evgmik replied:
fixed with 3962a45

evgmik committed:
added larger size in to the testing data set

sahib commented on fuse/fuse_test.go:
```diff
@@ -17,104 +19,290 @@ import (
-		t.Fatalf("Data differs over fuse: got %d, should be %d bytes", n, len(data))
+	mount := mountInfo{Dir: req.MntPath, Opts: req.Opts}
+	// function which required mounts
+	f(ctx, control, &mount)
```
More a FYI: When I use a variable only to pass it to a function, I usually pass it as literal, i.e:

```go
f(ctx, control, &mountInfo{
    Dir: req.MntPath,
    Opts: req.Opts
})
```

Just reduces my mental overhead when reading the code, since I can be sure that the variable is not used anywhere else and I don't need to check.

evgmik replied:
fixed in bb37469

sahib reviewed (changes requested):
Good work! All comments are more about style, less about functionality.

evgmik committed:
removed double check for the same condition

evgmik committed:
Spelling fix.

evgmik committed:
prepare a variable in the function call, instead of separate variable

evgmik committed:
Do unmount and cleanup with defer statement

Guarantees that we get to unmounting and cleaning, even
if function under test fails.

evgmik committed:
Better naming s/brig/catfs/ explicit and implicit

This avoid calling function with the same prefix as the main program (brig).
Also it uses proper terminology, since in brig the underlying file
system is catFS.

evgmik committed:
Better debug messages

evgmik committed:
using a helper function to compare catfs file content with etalon

evgmik committed:
More careful fuse write related check

File content match between fuse and catfs is not sufficient alone:
  file content might be miss-written.
So I added check for the correct content.

Additional checks for staging.
